### PR TITLE
PE-26 | Create home page

### PR DESCRIPTION
**Pull request by MaxLian11**
**What I did:**

- Created Home page demo with dummy records.
- I went a little beyond task PE-26: widgets Search and Category are functioning and are able to be tested using dummy records. Search bar can only search for a single keyword or better say a sequence of characters at this point.

**How to test:**

- In getQuestions function you can create different sets of dummy variables for different tests.
- Category filter must filter questions based on category, even if search was applied.
- Search bar must work properly even if category was changed in dropdown menu.

### CODE REVIEW

File: lib/category_filter.dart [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import 'package:flutter/material.dart';
+
+/// This is the stateful widget that the main application instantiates.
+class CategoryFilter extends StatefulWidget {
+  final _callback;
+
+  CategoryFilter({@required void getCategoryCallback(String category)}) : _callback = getCategoryCallback;
+
+  @override
+  _MyStatefulWidgetState createState() => _MyStatefulWidgetState();
```
Comment (tonynguyen98) [inline lib/category_filter.dart]: Can you rename `_MyStatefulWidgetState` to `_CategoryFilterState`

File: lib/category_filter.dart [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+/// This is the private State class that goes with MyStatefulWidget.
+class _MyStatefulWidgetState extends State<CategoryFilter> {
```
Comment (tonynguyen98) [inline lib/category_filter.dart]: Can you rename `_MyStatefulWidgetState` to `_CategoryFilterState`

File: lib/home_page.dart [hunk 3]
```diff
@@ -0,0 +1,314 @@
+import 'package:flutter/material.dart';
+import 'question_model.dart';
+import 'category_filter.dart';
+
+class Home extends StatefulWidget {
+  @override
+  HomeState createState() => HomeState();
+}
+
+class HomeState extends State<Home> {
```
Comment (tonynguyen98) [inline lib/home_page.dart]: Not as important, but can you rename `HomeState` to `_HomeState`

File: lib/home_page.dart [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+  // Might be inefficient to use that many lists around, thus, this might change in the future versions to something more dynamic and efficient
+  List questionsListFilteredCategory = [];
+  List questionsListFilteredSearch = [];
+  // isSearching keep track of whether search bar is activated or not. Search bar will appear if user clicked a search icon.
+  bool isSearching = false;
+  // Keeps track of the last chosen category. By default is All Categories.
+  String selectedCategory = "All Categories";
+
+  // Called from widget (class) category filter
+  getCategory(value) {
```
Comment (tonynguyen98) [inline lib/home_page.dart]: Please define the type to be a `void` here

File: lib/home_page.dart [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    setState(() {
+      filterQuestionsCategory(value);
+    });
+  }
+
+  // Here Eureka will get data from the database.
+  // At this point it contains just dummy lists of questions for test purposes
+  getQuestions() async {
```
Comment (tonynguyen98) [inline lib/home_page.dart]: Please define the type here to be `Future<void>`

File: lib/home_page.dart [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+
+  // Get data from the database to the list. Questions are shown on home page is always questionsListFiltered,
+  // which at this point is a copy of a `data`, as no filters have been applied yet
+  @override
+  void initState() {
+    getQuestions().then((data) {
+      setState(() {
+        questionsListFiltered = questionsListFilteredSearch = questionsListFilteredCategory = data;
+      });
+    });
```
Comment (tonynguyen98) [inline lib/home_page.dart]: Extract this function into it's own void function and then call it here.

File: lib/home_page.dart [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+            isSearching
+                ? IconButton(
+                    icon: Icon(Icons.cancel),
+                    onPressed: () {
+                      // Button to drop search bar. Will keep list filtered if category filter is applied
+                      setState(() {
+                        this.isSearching = false;
+                        questionsListFilteredSearch = data; // search list is now the same as the original
+                        filterQuestionsCategory(selectedCategory); // if "All Categories", then it page is completely unfiltered
+                      });
```
Comment (tonynguyen98) [inline lib/home_page.dart:165]: extract this out to be it's own function

File: lib/home_page.dart [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+                    },
```
Comment (tonynguyen98) [inline lib/home_page.dart]: extract the `onPressed` out to be it's own function

**Comment by tonynguyen98** [comment]
This looks really great! I have some requests about formatting changes. Ideally, we want to have each component on the screen to be it's own method. I haven't tested this yet but it looks really great!

File: lib/home_page.dart [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+                  textAlign: TextAlign.center,
+                ),
+              ),
+            ),
+          ),
+        ],
+      ),
+    );
+  }
+}
```
Comment (ayeejane) [inline lib/home_page.dart]: Button for "Create New Question" should not be visible when the user is typing on the search bar. You can extract this for now or comment it out.

**Review by ayeejane — CHANGES_REQUESTED** [review]
Tested it out on my end and everything works as intended. However, I think buttons for "Create New Question" and "Chat" can be extracted for now; it can be a separate issue to create those features in the near future. Also, there are a few formatting changes that Tony N. have already commented on but overall great job!

File: lib/home_page.dart (hunk 5 above)
Comment (MaxLian11) [reply lib/home_page.dart]: This caused errors on line 80: _Error: This expression has type 'void' and can't be used._
I tried `Future<dynamic>` instead and I think it worked. Will that be strongly typed?

Comment (tonynguyen98) [reply lib/home_page.dart]: Apologies, I didn't realize that you were returning something with this function. It looks like you're returning `List` so can you try if `Future<List>` will work? If not, then `Future<dynamic>` is fine.

File: lib/home_page.dart (hunk 9 above)
Comment (MaxLian11) [reply lib/home_page.dart]: Ok, I fixed that!
Why would you like to extract "new question" and "chat" buttons though? I think we can keep them.. They just don't have any functionalities yet

File: lib/home_page.dart (hunk 3 above)
Comment (MaxLian11) [reply lib/home_page.dart]: Sorry, forgot to change that, but we can do it later

File: lib/home_page.dart (hunk 9 above)
Comment (ayeejane) [reply lib/home_page.dart]: I guess you could leave it. I just thought it could be a separate issue.

**Review by ayeejane — APPROVED** [review]

**Review by txt020 — APPROVED** [review]
Everything looks good!!

File: lib/home_page.dart (hunk 3 above)
Comment (tonynguyen98) [reply lib/home_page.dart]: Please fix this when you fix any merge conflicts.

File: lib/home_page.dart [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+            // Chat button
+            FlatButton(
+              textColor: Colors.white,
+              onPressed: () {}, // redirect to Chat page
+              child: Icon(Icons.chat_sharp, color: Colors.cyan),
+              shape: CircleBorder(side: BorderSide(color: Colors.transparent)),
+            ),
+          ],
+          backgroundColor: Colors.blueGrey[800],
+        ),
```
Comment (tonynguyen98) [inline lib/home_page.dart]: Extract this outside of the `Widget build` something like this: 
```dart
AppBar homeAppBar() {
return AppBar(
          // Menu button
          leading: FlatButton(
            textColor: Colors.white,
            onPressed: () {}, // open side menu
            child: Icon(Icons.menu, color: Colors.cyan),
            shape: CircleBorder(side: BorderSide(color: Colors.transparent)),
          ),

          // Title "Eureka" is changed to a search bar after search icon is clicked
          title: !isSearching
              ? Text('Eureka!')
              : TextField(
                  onChanged: (value) {
                    // Call function to apply search filter
                    filterQuestionsSearch(value);
                  },
                  style: TextStyle(color: Colors.white),
                  decoration: InputDecoration(
                      icon: Icon(
                        Icons.search,
                        color: Colors.white,
                      ),
                      hintText: "Search Question",
                      hintStyle: TextStyle(color: Colors.white)),
                ),
          centerTitle: true,

          // Search icon is changed to search bar
          actions: <Widget>[
            isSearching
                ? IconButton(
                    icon: Icon(Icons.cancel),
                    onPressed: () {
                      searchingFalse();
                    },
                  )
                : IconButton(
                    icon: Icon(Icons.search),
                    onPressed: () {
                     searchingTrue();
                    },
                  ),

            // Chat button
            FlatButton(
              textColor: Colors.white,
              onPressed: () {}, // redirect to Chat page
              child: Icon(Icons.chat_sharp, color: Colors.cyan),
              shape: CircleBorder(side: BorderSide(color: Colors.transparent)),
            ),
          ],
          backgroundColor: Colors.blueGrey[800],
        );
}

Comment (tonynguyen98) [reply lib/home_page.dart]: And then reference `homeAppBar()` instead. We don't want such bug chunks of code inside the `Widget build`.

File: lib/home_page.dart [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+                  "Create New Question",
+                  maxLines: 1,
+                  style: TextStyle(color: Colors.white),
+                  textAlign: TextAlign.center,
+                ),
+              ),
+            ),
+          ),
+        ],
+      ),
```
Comment (tonynguyen98) [inline lib/home_page.dart]: Similarly here:
```dart
Column questionsList(){
return Column(
        children: [
          // List of Questions
          Expanded(
              // Show loading circle if results are taking time
              // Show "No results" if input text doesn't match with question title (later will be added to description too)
              child: _buildList(questionsListFiltered, filterQuestionsCategory, getCategory)),

          RawMaterialButton(
            onPressed: () {},
            fillColor: Colors.blueGrey[800],
            splashColor: Colors.grey,
            child: Padding(
              padding: EdgeInsets.all(10.0),
              child: SizedBox(
                width: 200.0,
                child: Text(
                  "Create New Question",
                  maxLines: 1,
                  style: TextStyle(color: Colors.white),
                  textAlign: TextAlign.center,
                ),
              ),
            ),
          ),
        ],
      );
}
```

File: lib/home_page.dart [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+                        textAlign: TextAlign.left,
+                        style: TextStyle(fontSize: 15.0),
+                      ),
+                      SizedBox(height: 5),
+                      Row(children: [
+                        Text(
+                          "Asked:  ",
+                          textAlign: TextAlign.left,
+                          style: TextStyle(fontSize: 15.0, color: Colors.grey),
+                        ),
```
Comment (tonynguyen98) [inline lib/home_page.dart]: Here, you kinda repeat the similar code. Please make a custom Text that you can pass variables into and allow for cleaner code. Example: 
```dart
Text customText (String string, int fontSize, Colors color) {
return Text(
                          string,
                          textAlign: TextAlign.left,
                          style: TextStyle(fontSize: fontSize, color: color),
                        ),
}
```
Then you call `customText ("Asked:  ", 15.0, Colors.grey)`

**Review by tonynguyen98 — CHANGES_REQUESTED** [review]
Some changes to provide for cleaner code.

**Comment by tonynguyen98** [comment]
Also, please resolve conservations when you complete them.

**Review by tonynguyen98 — APPROVED** [review]
Great work on getting your first flutter screen done! Great job!!

**Pull request merged**